pmlc/conversion/tile_to_pxa/tile_to_pxa.cc

```diff
@@ -890,46 +891,225 @@ struct GatherOpConversion : public OpConversionPattern<tile::GatherOp> {
+    for (size_t i = 0; i < dstDims; ++i) {
       srcOps.push_back(loop.getIVs()[i]);
     }
 
-    // load the specified value from the source tensor
-    auto loaded = rewriter.create<mlir::LoadOp>(loc, tensor, srcOps);
+    // create std ops for 1D interpolation
+    Value interpVal;
+    if (index.getType().isa<FloatType>()) {
+      switch (op.mode()) {
+      case InterpolationMode::linear:
+        interpVal = buildLinearInterpolationOps(loc, rewriter, tensor, index,
+                                                srcOps, dim);
+        break;
+      case InterpolationMode::cubic:
+        interpVal = buildCubicInterpolationOps(
+            loc, rewriter, tensor, index, srcOps, dim, op.cubicCoeffAttr());
+        break;
+      case InterpolationMode::nearest:
+      default:
+        interpVal = buildNearestInterpolationOps(loc, rewriter, tensor, index,
+                                                 srcOps, dim);
+        break;
+      }
+    } else {
+      if (!index.getType().isa<IndexType>()) {
+        auto indexType = rewriter.getIndexType();
+        // cast from whatever integer type it has to index type
+        index = rewriter.create<mlir::IndexCastOp>(loc, index, indexType)
+                    .getResult();
+      }
+      srcOps.at(dim) = index;
+      interpVal = rewriter.create<mlir::LoadOp>(loc, tensor, srcOps);
+    }
 
     // create a destination map using all of the dimensions
     auto dstStoreMap = AffineMap::getMultiDimIdentityMap(dstDims, ctx);
 
     // create a destination map from the whole loop
     auto stored = rewriter.create<pxa::PxaReduceOp>(loc, AtomicRMWKind::assign,
-                                                    loaded, resultMemRef,
+                                                    interpVal, resultMemRef,
                                                     dstStoreMap, loop.getIVs());
     rewriter.create<AffineYieldOp>(loc, ArrayRef<Value>{stored.getResult()});
 
     rewriter.replaceOp(op, loop.getResult(0));
 
     return success();
   }
+
+  Value buildNearestInterpolationOps(Location loc,
+                                     ConversionPatternRewriter &rewriter,
+                                     Value tensor, Value index,
+                                     std::vector<Value> &srcOps,
+                                     int dim) const {
+    auto f32Type = rewriter.getF32Type();
+    auto i32Type = rewriter.getI32Type();
+    auto indexType = rewriter.getIndexType();
+
+    auto cstHalfF32 = rewriter.create<mlir::ConstantOp>(
+        loc, f32Type, rewriter.getFloatAttr(f32Type, 0.5));
+    index = rewriter.create<mlir::AddFOp>(loc, index, cstHalfF32);
+    index = rewriter.create<mlir::FPToUIOp>(loc, index, i32Type).getResult();
+    index =
+        rewriter.create<mlir::IndexCastOp>(loc, index, indexType).getResult();
+    srcOps.at(dim) = index;
+    return rewriter.create<mlir::LoadOp>(loc, tensor, srcOps);
+  }
+
+  Value buildLinearInterpolationOps(Location loc,
```

Comment:
Hmm, I wonder if we could write `.mlir` files instead and just link in these functions instead of generating them here. It's a bit hard to read in C++ format, but would be much cleaner in a `.mlir` file.

I don't think we need to modify this PR, but in the future, I'd like to explore how we can write complex code like this in a more maintainable fashion.

Reply:
Yes, that would be much more mantainable.